### ISSUE TITLE
tests are not cleaning up projects/namespaces.

### ISSUE THREAD

**Issue by kadel** [openshift/odo#4467]
Integrations and e2e tests are not properly cleaning up project/namespaces.
Currently, there are 77 projects with random 10 letter name (those are created by our integration tests)

```
kubectl get projects -o=custom-columns=NAME:.metadata.name | grep -E "^[a-z]{10}$" | wc -l
77
```
The above issue is tracked with [issue 4502](https://github.com/openshift/odo/issues/4502)

On top of that, we have 90 projects starting with `odo-pr-build`
```
kubectl get projects -o=custom-columns=NAME:.metadata.name | grep -E "odo-pr-build" | wc -l
90
````

Most of those project are older than 1 day (`kubectl get projects -o=custom-columns=NAME:.metadata.name,CREATED:.metadata.creationTimestamp`) so they are not from currently running tests .

**Comment by prietyc123** [comment]
@kadel I think it is cluster specific because I tried running running whole test suite 2-3 times and captures the logs into a file, despite of test failures I didn't find garbage projects left over.

```
$ oc version
Client Version: 4.4.0-0.ci-2020-02-26-215902
Server Version: 4.7.0-rc.3
Kubernetes Version: v1.20.0+bd9e442

$ oc get projects -o=custom-columns=NAME:.metadata.name | grep -E "^[a-z]{10}$" | wc -l
       0
```
Test logs:
[log3.txt](https://github.com/openshift/odo/files/6081803/log3.txt)
[log4.txt](https://github.com/openshift/odo/files/6081804/log4.txt)

Further I will try researching on PSI cluster as mentioned above.

**Comment by rnapoles-rh** [comment]
With PR https://github.com/openshift/odo/pull/4509 we will use not fully random project names, this will provide some light if we see projects left behind

**Comment by rnapoles-rh** [comment]
Blocked by 4.5 cluster not functional. I will get back to this once #4515 is resolved

**Comment by rnapoles-rh** [comment]
When running tests locally in minishift I noticed a lot of project left behind despite they claim to be deleted:
```
Deleting project: cmd-push-test138dem
Running odo with args [odo project delete cmd-push-test138dem -f]
[odo] I0409 15:04:44.557187   17174 util.go:727] HTTPGetRequest: https://raw.githubusercontent.com/openshift/odo/master/build/VERSION
[odo] I0409 15:04:44.557974   17174 util.go:748] Response will be cached in /var/folders/1m/ynw0pw815hq2fmy5zrhktpyh0000gn/T/odohttpcache for 1h0m0s
[odo] I0409 15:04:44.558199   17174 util.go:761] Cached response used.
[odo] I0409 15:04:44.588855   17174 kclient.go:202] Checking if "projects" resource supported
[odo] I0409 15:04:44.612075   17174 kclient.go:202] Checking if "projects" resource supported
[odo]  ✓  Deleted project : cmd-push-test138dem
[odo]  ⚠  Warning! Projects are deleted from the cluster asynchronously. Odo does its best to delete the project. Due to multi-tenant clusters, the project may still exist on a different node.
[odo] I0409 15:04:44.621816   17174 odo.go:72] Could not get the latest release information in time. Never mind, exiting gracefully :)
```
If I run oc projects after the tests completed I can see that cmd-push-test138dem is listed

**Comment by prietyc123** [comment]
> When running tests locally in minishift I noticed a lot of project left behind despite they claim to be deleted

@rnapoles-rh If I understand it correctly, projects might not be getting cleaned even after performing cleanup steps in `AfterEach`. Object might have stuck in  `Terminating` state and waiting on finalizer. If this is the case, only thing we can think of for QE progress is removing it forcefully. Recently I was going through some related documents and found some workaround to remove finalizer thing. AFAIK this is a known bug and there is no sol of it. 
However I was thinking aloud to implement those workaround either in the backend code or in `AfterEach` test step to unblock QE further psi implementation. 
Due to improper clean up/Out of memory, the cluster becomes slow and server takes long time to respond which results further odo steps to fail. For ex:
```
Running odo.exe with args [odo project create e2e-images-test114csf -w -v4]

[ssh:Windows 10] [2] [odo] I0421 11:31:33.007343    5984 util.go:761] Cached response used.

[ssh:Windows 10] [2] [odo] I0421 11:31:32.973130    5984 preference.go:217] The path for preference file is C:\Users\Admin\886927573\preference.yaml

[ssh:Windows 10] [2] [odo]  -  Waiting for project to come up  ...

[ssh:Windows 10] [2] [odo] I0421 11:31:54.028015    5984 kclient.go:202] Checking if "projects" resource supported

[ssh:Windows 10] [2] [odo] 
 X  Waiting for project to come up [21s]

[ssh:Windows 10] [2] [odo]  X  Get "https://api.testocp47.psiodo.net:6443/apis/project.openshift.io/v1?timeout=32s": dial tcp 10.0.108.16:6443: connectex: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.
```
Earlier we were facing same issue on travis as well. There are some discussion around it https://github.com/openshift/odo/issues/4426#issuecomment-779809624 . IMO meanwhile we get fix for `odo project delete` which is affecting other scenarios to fail as well, let's proceed with QE [workaround to remove projects forcefully](https://stackoverflow.com/questions/58638297/project-deletion-struck-in-terminating). There are number of approaches discussed, we can move forward with the best fit in sol.
@kadel WDYT?

**Comment by prietyc123** [comment]
> When running tests locally in minishift I noticed a lot of project left behind despite they claim to be deleted

@rnapoles-rh If I understand it correctly, projects might not be getting cleaned even after performing cleanup steps in `AfterEach`. Object might have stuck in  `Terminating` state and waiting on finalizer. If this is the case, only thing we can think of for QE progress is removing it forcefully. Recently I was going through some related documents and found some workaround to remove finalizer thing. AFAIK this is a known bug and there is no sol of it. 
However I was thinking aloud to implement those workaround either in the backend code or in `AfterEach` test step to unblock QE further psi implementation. 
Due to improper clean up/Out of memory, the cluster becomes slow and server takes long time to respond which results further odo steps to fail. For ex:
```
Running odo.exe with args [odo project create e2e-images-test114csf -w -v4]

[ssh:Windows 10] [2] [odo] I0421 11:31:33.007343    5984 util.go:761] Cached response used.

[ssh:Windows 10] [2] [odo] I0421 11:31:32.973130    5984 preference.go:217] The path for preference file is C:\Users\Admin\886927573\preference.yaml

[ssh:Windows 10] [2] [odo]  -  Waiting for project to come up  ...

[ssh:Windows 10] [2] [odo] I0421 11:31:54.028015    5984 kclient.go:202] Checking if "projects" resource supported

[ssh:Windows 10] [2] [odo] 
 X  Waiting for project to come up [21s]

[ssh:Windows 10] [2] [odo]  X  Get "https://<psi-cluster-url>/apis/project.openshift.io/v1?timeout=32s": dial tcp <psi-ip><port>: connectex: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.
```
Earlier we were facing same issue on travis as well. There are some discussion around it https://github.com/openshift/odo/issues/4426#issuecomment-779809624 . IMO meanwhile we get fix for `odo project delete` which is affecting other scenarios to fail as well, let's proceed with QE [workaround to remove projects forcefully](https://stackoverflow.com/questions/58638297/project-deletion-struck-in-terminating). There are number of approaches discussed, we can move forward with the best fit in sol.
@kadel WDYT?

**Comment by kadel** [comment]
Before we start with a workaround let's try to understand why is this happening.
If the project is not deleted even after it is marked  for deletion it meant that there have to be some resources left in that project that is waiting for something (pvc cleanup, finalizers etc...) If we try to understand the problem first we might be able to fix it properly instead of just implementing a workaround that will bite us back in the future.

**Comment by rnapoles-rh** [comment]
Adding code in AfterEach to track each passing/failing test, including project name to assist in analysis.

**Comment by anandrkskd** [comment]
Scope for this issue
1. For project created by odo with random name
- Create a draft PR for debug purpose, will follow [URL](https://www.openshift.com/blog/the-hidden-dangers-of-terminating-namespaces) contains some process that can help debug if any resources are interfering with the release process. 
2. For projects that are created by test bash script for tests and not get deleted due to abort or other reason, example project name `odo-pr-build*`
- Create a post build script that checks and delete if those are left behind after the test completes for a particular PR.

**Comment by anandrkskd** [comment]
1. For project created by odo with random name
- In recent builds no Project created by our test was left behind, this was verified by running test using jenkins and manually.

2. For projects that are created by test bash script for tests and not get deleted due to abort or other reason, example project name odo-pr-build*
-  PR is created in internal repository to recheck in post build if we have any project that is left after each test is executed.

**Comment by kadel** [comment]
> 1. For project created by odo with random name
> 
> * In recent builds no Project created by our test was left behind, this was verified by running test using jenkins and manually.
> 




> 1. For projects that are created by test bash script for tests and not get deleted due to abort or other reason, example project name odo-pr-build*
> 
> * PR is created in internal repository to recheck in post build if we have any project that is left after each test is executed.


I can still see a lot of projects, some are from May
```
▶ oc get projects -o jsonpath='{range .items[*]}{.metadata.name} {"\t"} {.metadata.creationTimestamp} {"\n"}{end}' | grep pr-build

odo-mac-e2e-pr-buildpr4658hy4f 	 2021-06-21T04:45:35Z 
odo-mac-e2e-pr-buildpr4658kf2a 	 2021-06-21T04:45:33Z 
odo-mac-e2e-pr-buildpr4658lsff 	 2021-06-21T04:45:36Z 
odo-mac-e2e-pr-buildpr4658uaxh 	 2021-06-21T04:45:38Z 
odo-mac-e2e-pr-buildpr4658yh8i 	 2021-06-21T04:45:40Z 
odo-mac-e2e-pr-buildpr49462pmw 	 2021-07-28T07:11:20Z 
odo-mac-e2e-pr-buildpr494663uk 	 2021-07-28T07:11:24Z 
odo-mac-e2e-pr-buildpr4946q0ox 	 2021-07-28T07:11:21Z 
odo-mac-e2e-pr-buildpr4946y1hv 	 2021-07-28T07:11:23Z 
odo-mac-e2e-pr-buildpr4946y392 	 2021-07-28T07:11:18Z 
odo-pr-buildpr46584he2 	 2021-05-24T05:26:54Z 
odo-pr-buildpr4658b9po 	 2021-05-24T05:26:56Z 
odo-pr-buildpr4658fd53 	 2021-05-24T05:26:55Z 
odo-pr-buildpr4658g4hv 	 2021-05-24T05:26:58Z 
odo-pr-buildpr4658oqk3 	 2021-05-24T05:26:59Z 
odo-pr-buildpr4681hj56 	 2021-05-04T08:26:25Z 
odo-pr-buildpr4681lwwh 	 2021-05-04T08:28:22Z 
odo-pr-buildpr4681tn4d 	 2021-05-04T08:26:23Z 
odo-pr-buildpr4681tr1c 	 2021-05-04T08:28:25Z 
odo-pr-buildpr4681x5yg 	 2021-05-04T08:26:27Z 
odo-pr-buildpr4689dsw6 	 2021-05-05T09:01:00Z 
odo-pr-buildpr4689h0uf 	 2021-05-05T09:00:58Z 
odo-pr-buildpr4689qa7h 	 2021-05-05T09:00:57Z 
odo-pr-buildpr4689scty 	 2021-05-05T09:00:56Z 
odo-pr-buildpr4689vhwm 	 2021-05-05T09:01:01Z 
odo-pr-buildpr47019okj 	 2021-05-11T07:13:12Z 
odo-pr-buildpr4701gfgb 	 2021-05-11T07:13:08Z 
odo-pr-buildpr4701j0pl 	 2021-05-11T07:13:10Z 
odo-pr-buildpr4701l87t 	 2021-05-11T07:13:14Z 
odo-pr-buildpr4701nkxj 	 2021-05-11T07:13:11Z 
odo-pr-buildpr4748hftl 	 2021-06-03T09:13:04Z 
odo-pr-buildpr4748m912 	 2021-06-03T09:13:02Z 
odo-pr-buildpr4748p2df 	 2021-06-03T09:13:05Z 
odo-pr-buildpr4748vdmy 	 2021-06-03T09:13:06Z 
odo-pr-buildpr4748zjqh 	 2021-06-03T09:13:03Z 
odo-windows-e2e-pr-buildpr46583bww 	 2021-06-10T12:52:39Z 
odo-windows-e2e-pr-buildpr46587y19 	 2021-06-10T12:59:48Z 
odo-windows-e2e-pr-buildpr46589jhs 	 2021-06-10T12:56:08Z 
odo-windows-e2e-pr-buildpr4658axyd 	 2021-06-10T13:06:10Z 
odo-windows-e2e-pr-buildpr4658o03v 	 2021-06-10T13:03:01Z 
odo-windows-e2e-pr-buildpr494636l0 	 2021-07-27T08:58:14Z 
odo-windows-e2e-pr-buildpr4946gzwo 	 2021-07-27T08:58:19Z 
odo-windows-e2e-pr-buildpr4946mvrf 	 2021-07-27T08:58:17Z 
odo-windows-e2e-pr-buildpr4946rsga 	 2021-07-27T08:58:16Z 
odo-windows-e2e-pr-buildpr4946wk4k 	 2021-07-27T08:58:21Z 
```

**Comment by anandrkskd** [comment]
For older PRs we need to remove it manually,  but the new projects that were not deleted was because failure in my post-build cleanup script.

**Comment by dharmit** [comment]
@feloy can you describe the problem you're experiencing while running tests on CRC?

**Comment by feloy** [comment]
I'm not sure it is related to this issue: when I run tests locally on CRC and stops the tests manually for some reason, the projects/namespaces are not removed (this is normal as AfterEach didn't get the chance to be executed). But more than the namespaces, the resources (pods, etc) declared into these namespaces are not removed either, which gets a lot of memory resources (also normal, because we rely on the namespace deletion to delete their resources).

My question is: are there some resources still present in the remaining namespaces? If so, it could help understand why the namespaces are not deleted.

**Comment by mohammedzee1000** [comment]
> I'm not sure it is related to this issue: when I run tests locally on CRC and stops the tests manually for some reason, the projects/namespaces are not removed (this is normal as AfterEach didn't get the chance to be executed). But more than the namespaces, the resources (pods, etc) declared into these namespaces are not removed either, which gets a lot of memory resources (also normal, because we rely on the namespace deletion to delete their resources).
> 
> My question is: are there some resources still present in the remaining namespaces? If so, it could help understand why the namespaces are not deleted.

Yes, and unfortunately there is nothing we can do about that. Preemption of tests for any reason(ctrl+c, jenkins slave dies, ssh connection breaks etc) is going to cause this problem

**Comment by dharmit** [comment]
I changed the priority of this to urgent because without this problem being fixed, moving to IBM Cloud for CI won't magically make our CI more reliable. Even addressing #3256 boiled down to fixing this issue.

**Comment by mohammedzee1000** [comment]
> > I'm not sure it is related to this issue: when I run tests locally on CRC and stops the tests manually for some reason, the projects/namespaces are not removed (this is normal as AfterEach didn't get the chance to be executed). But more than the namespaces, the resources (pods, etc) declared into these namespaces are not removed either, which gets a lot of memory resources (also normal, because we rely on the namespace deletion to delete their resources).
> > My question is: are there some resources still present in the remaining namespaces? If so, it could help understand why the namespaces are not deleted.
> 
> Yes, and unfortunately there is nothing we can do about that. Preemption of tests for any reason(ctrl+c, jenkins slave dies, ssh connection breaks etc) is going to cause this problem

Considering this, I think we are going to need a 2 pronged approach to solve this problem.
1. Ensure that unless there is pre-emption all projects get removed
2. Setup a script to allow cleanup of namespaces/projects.

For random projects created in setup script right now, it might make sense to move that logic into the test files themselves so that we can control their deletion. This will of course need the tests to allow cleanup like all other tests. The thing the keep in mind though is these projects are outside because they need some kind of kubeadmin access. So helpers might need to be added to allow the actions to temporarily happen as kubeadmin before switching back original user.

The script should be able to ignore Kubernetes/openshift infra projects and delete the projects and namespace based on creation timestamp (older than maybe).
This should besimple enough for oc using `oc get projects -o jsonpath='{range .items[*]}{.metadata.name} {"\t"} {.metadata.creationTimestamp} {"\n"}{end}'` as we do timestamps that we can use but we will need to investigate further for kubectl namespaces

**Comment by kadel** [comment]
> The script should be able to ignore Kubernetes/openshift infra projects and delete the projects and namespace based on creation timestamp (older than maybe).

`namespace`/`project` is k8s resource like any other. So you can attach labels to them and then use label selectors to delete.

Only downside that our scripts won't be able to use `odo project create`, and all projects and namespace will have to be created using `oc create` or `kubectl create`.

**Comment by anandrkskd** [comment]
> delete the projects and namespace based on creation timestamp (older than maybe).

Won't that be problem if there we are using concurrent builds? 
Instead can we use name created by joining PR numbers  and random string, and use that to delete everytime we before we exit?

**Comment by mohammedzee1000** [comment]
> > delete the projects and namespace based on creation timestamp (older than maybe).
> 
> Won't that be problem if there we are using concurrent builds?
> Instead can we use name created by joining PR numbers and random string, and use that to delete everytime we before we exit?

You cannot control preemptions

**Comment by anandrkskd** [comment]
> You cannot control preemptions

I am not able to get what scenario you are thinking of, can you share in what kind of scenario it will be a problem ?

**Comment by mohammedzee1000** [comment]
> > You cannot control preemptions
> 
> I am not able to get what scenario you are thinking of, can you share in what kind of scenario it will be a problem ?

Any senaio that causes primitive exit. OOM kill for eg if the Jenkins slave run out of memory or ssh connection breaking due to network issue or even someone doing. ctrl+c during execution.

If the ginkgo process gets killed, then it will never get to cleanup even if you have the best cleanup of the code.

It is one of many reasons prow gives fresh cluster for every run. So even if test does in middle cleanup is ensured by just destroying entire cluster.

When this happens and believe me it will, the cleanup won't happen.

There is no easy way to track the projects being left behind so additional cleanup will be needed based on something we can track. 

For example we should not have projects older than maybe a week in a resting cluster.

Ofcourse infra projects used by system which there are quite a few, should not be cleaned up.

**Comment by anandrkskd** [comment]
Understood, 
what I understood from this [comment](https://github.com/openshift/odo/issues/4467#issuecomment-888897564) is that the setup of e2e script will also need to be done using ginkgo with helper function which require kubeadmin access? 
and a script that will clean up namespaces/projects that are more then `X` days  old? right?
If so what is the no. of days we are thinking off?

**Comment by mohammedzee1000** [comment]
> Understood,
> what I understood from this [comment](https://github.com/openshift/odo/issues/4467#issuecomment-888897564) is that the setup of e2e script will also need to be done using ginkgo with helper function which requires kubeadmin access?

Yes, if we can do that, then cleanup can be done the same way we do for other tests, Right now, we have to do manual cleanup logging in as kubeadmin

We could do something like
```go
helper.AsAdmin(func() {
})

```
the helper would
1. Kubeconfig isolation. CommonBeforeEach already does this so we hould be ok here
2. Check if Openshift cluster (we are usually kubeadmin on Kubernetes. If we use dev account for Kubernetes, then we will need to figure out how to switch accounts there too)
For openshift
```bash
$ oc whoami -t # gives token of current user which we should save
$ oc login -u kubeadmin -p kubeamdinpass # or oc login --token=kubeadmin token
# do the dirty stuff that needs kubeadmin
$ oc login --token=savedtokenfromwhoami
```
Put simply the first 2 commands will come before calling the function and last command at the end. Kubeconfig is cleaned up as part of CommonAfterEach so we doing this much should be enough
```
> and a script that will clean up namespaces/projects that are more then `X` days old? right?
> If so what is the no. of days we are thinking off?

That is something we should consider. I personally feel any project older than a day should not need to exist (unless its infra project / default/ myproject ie first project). But we can expand that to be safe as we don't know how long someone else's test will run

**Comment by anandrkskd** [comment]
> That is something we should consider. I personally feel any project older than a day should not need to exist (unless its infra project / default/ myproject ie first project). But we can expand that to be safe as we don't know how long someone else's test will run

so can't we use a prefix that will help identify our project and only delete Those? we will anyway have a namespace with random text we will add a prefix to avoid deletion of projects that are not created by us/our tests.

**Comment by mohammedzee1000** [comment]
> > That is something we should consider. I personally feel any project older than a day should not need to exist (unless its infra project / default/ myproject ie first project). But we can expand that to be safe as we don't know how long someone else's test will run
> 
> so can't we use a prefix that will help identify our project and only delete Those? we will anyway have a namespace with random text we will add a prefix to avoid deletion of projects that are not created by us/our tests.

We can do that, yes.B but I think an overall strat would be a better way to go about especially considering we will be sharing. You can have one solution that should work for everyone who shares the cluster as it is a cluster level problem that all of us will face

**Comment by anandrkskd** [comment]
In that case I think we will need a file that contains list of project that we don't want to delete? and other team will require to add there project that they don't want to delete?

**Comment by dharmit** [comment]
> In that case I think we will need a file that contains list of project that we don't want to delete? and other team will require to add there project that they don't want to delete?

We only need to fetch a list of projects older than X amount of time and that doesn't start with `openshift`, `kubernetes` or `default`. All such project should be deleted. A fresh CRC cluster had zero projects that didn't start with either of these three.

This can be done using `kubectl get ns` (it has an `AGE` column).

**Comment by dharmit** [comment]
> We only need to fetch a list of projects older than X amount of time and that doesn't start with `openshift`, `kubernetes` or `default`. All such project should be deleted. 

Saying this based on the "NOTE" at the top of [OpenShift 4.8 docs](https://docs.openshift.com/container-platform/4.8/applications/projects/working-with-projects.html).

**Comment by dharmit** [comment]
Run the job as a k8s cron job everyday:
- https://v1-19.docs.kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/
- https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/

X to be 1 days instead of 2.

**Comment by anandrkskd** [comment]
As per discussion with @mohammedzee1000 and @prietyc123, This issue will require changes in ginkgo function to add of generic function for openshift as well as kubernetes tests, changes in test of Postgres operator and for e2e test adding secrets.
